### PR TITLE
Bring MPC to the new flight stack

### PR DESCRIPTION
**Pull request by foehnph**
I made some changes to allow compatibility with our new MPC.
It includes mainly templating the autopilot on the controller type and parameter type, which requires the autopilot.cpp to move to the header file, solved as a head include.

Further changes were also mentioned in #5 and we concluded that references are always given in the form of a `Trajectory` (and no longer `TrajectoryPoint`) to the controller.
The position controller would then just take the first point in it, while the MPC can use the first (for a single reference) or a whole lookahead on a trajectory.

I made several changes and would be happy to hear your comments.

@internals: also take a look at: https://github.com/uzh-rpg/rpg_mpc/pull/5

### CODE REVIEW

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 1]
```diff
@@ -857,7 +878,7 @@ quadrotor_common::ControlCommand AutoPilot::start(
   }
 
   if (timeInCurrentState() > optitrack_start_land_timeout_
-      || reference_state_.position.z() >= optitrack_start_height_)
+      || state_estimate.position.z() >= optitrack_start_height_)
```
Comment (kohlerj) [inline control/autopilot/include/autopilot/autopilot_inl.h]: why the change to `state_estimate` from `reference_state` ?

Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: Whats the intention behind checking what the reference height is? Doesn't it make more sense to check what your actual height is and cancel the take-off once you reached this height?

Comment (kohlerj) [reply control/autopilot/include/autopilot/autopilot_inl.h]: if you look further in the function the reference is moved up and up as the loop is repeated. I would say that using the old method makes it more robust and smoother but this is only a feeling. By that I mean, hover WILL be reached even if you have an offset for idk which reason.

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 2]
```diff
@@ -1085,13 +1141,20 @@ quadrotor_common::ControlCommand AutoPilot::velocityControl(
 
   time_last_velocity_command_handled_ = ros::Time::now();
 
+
+  reference_trajectory_ = quadrotor_common::Trajectory();
+  reference_trajectory_.trajectory_type = 
+    quadrotor_common::Trajectory::TrajectoryType::GENERAL;
```
Comment (kohlerj) [inline control/autopilot/include/autopilot/autopilot_inl.h]: quick question, how about a constructor with the type as argument for the class `quadrotor_common::Trajectory()` ? to avoid code duplication. We could even have a constructor with one point as argument and automatically defined it as a general trajectory. It would make sense since we change the interface completely (i mean from point to trajectory in the run command...)

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 3]
```diff
@@ -1107,13 +1170,20 @@ quadrotor_common::ControlCommand AutoPilot::followReference(
...
 
   return command;
 }
 
-quadrotor_common::ControlCommand AutoPilot::executeTrajectory(
+template <typename Tcontroller, typename Tparams>
+quadrotor_common::ControlCommand
+AutoPilot<Tcontroller, Tparams>::executeTrajectory(
     const quadrotor_common::QuadStateEstimate& state_estimate,
     ros::Duration* trajectory_execution_left_duration,
```
Comment (kohlerj) [inline control/autopilot/include/autopilot/autopilot_inl.h]: general comment: really a point to a ros::Duration and int ?!? ---> how about a smart pointer?

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 4]
```diff
@@ -1147,20 +1222,57 @@ quadrotor_common::ControlCommand AutoPilot::executeTrajectory(
...
 
   const ros::Duration dt = time_now - time_start_trajectory_execution_;
   reference_state_ = trajectory_queue_.front().getStateAtTime(dt);
+  reference_trajectory_ = quadrotor_common::Trajectory();
+  reference_trajectory_.trajectory_type = 
+    quadrotor_common::Trajectory::TrajectoryType::GENERAL;
+
+  std::list<quadrotor_common::TrajectoryPoint>::const_iterator it_points;
+  std::list<quadrotor_common::Trajectory>::const_iterator it_trajectories;
+  bool lookahead_reached(false);
```
Comment (kohlerj) [inline control/autopilot/include/autopilot/autopilot_inl.h]: - could you comment the code, basic idea behind the heuristic (i'm lazy but so will be the next person working on it)
- should it not be a separate function like mpc?
- could we use specialization for these functions that change a lot ? I mean a definition for execute(mpc controller templated) and for execute(normal controller) ... Idk if it's possible but I think so. [](http://en.cppreference.com/w/cpp/language/template_specialization
)

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 2 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: Agree! Will do it tomorrow!

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 3 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: meh... lets make it pass-by-reference!
then there's no pointer magic but still avoiding copying, ok?

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 4 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: 3rd bullet: nope... if so we need to specify the template type in here... then we need the type at compile time.
but this is what we want to avoid, to be able to compile rpg_quadrotor_control without rpg_mpc.
I will think about making it more readable and understandable.

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 1 above)
Comment (mfaessle) [reply control/autopilot/include/autopilot/autopilot_inl.h]: So ideally the drone does what the reference tells it to, and in that case it does not make much of a difference. However, if e.g. the drone has a wrong thrust mapping and flies much lower than the reference, when using the state estimate, it will go up until it reached the start height (with a reference much higher) but then when it switches back to hover the reference is set to the current state estimate and the drone will drop by that much again which also does not result in the desired height. So by checking the reference height at least we get a smooth behavior in this case. Anyhow, if the drone does not follow the reference well it needs to be fixed...
Bottom line: I would change this back to reference state

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 2 above)
Comment (mfaessle) [reply control/autopilot/include/autopilot/autopilot_inl.h]: also agree ;)

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 3 above)
Comment (mfaessle) [reply control/autopilot/include/autopilot/autopilot_inl.h]: No, it is our convention that arguments which are changed by the function are passed as pointers.

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 5]
```diff
@@ -1147,20 +1222,57 @@ quadrotor_common::ControlCommand AutoPilot::executeTrajectory(
       *trajectory_execution_left_duration = ros::Duration(0.0);
       *trajectories_left_in_queue = 0;
       trajectory_queue_.pop_front();
-      setAutoPilotState(States::HOVER);
-      return base_controller_.run(state_estimate, reference_state_,
+      setAutoPilotStateForced(States::HOVER);
```
Comment (mfaessle) [inline control/autopilot/include/autopilot/autopilot_inl.h]: This should not be forced to avoid jumpy behavior if for whatever reason the trajectory ends with a non zero velocity. So by not forcing it will do a smooth breaking if necessary

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 6]
```diff
@@ -1147,20 +1222,57 @@ quadrotor_common::ControlCommand AutoPilot::executeTrajectory(
...
+        quadrotor_common::Trajectory::TrajectoryType::GENERAL;
+      reference_trajectory_.points.push_back(reference_state_);
+      return base_controller_.run(state_estimate, reference_trajectory_,
                                   base_controller_params_);
     }
     else
     {
       time_start_trajectory_execution_ +=
-          trajectory_queue_.front().points.back().time_from_start;
+        trajectory_queue_.front().points.back().time_from_start;
```
Comment (mfaessle) [inline control/autopilot/include/autopilot/autopilot_inl.h]: I saw a few of those... Detail but this is just a thing because of non equal auto formatting which doesn't hurt but is also not necessary.

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 4 above)
Comment (mfaessle) [reply control/autopilot/include/autopilot/autopilot_inl.h]: I will look at this in more detail later then

File: control/autopilot/include/autopilot/autopilot.h [hunk 7]
```diff
@@ -137,10 +138,11 @@ class AutoPilot
 
   state_predictor::StatePredictor state_predictor_;
 
-  position_controller::PositionController base_controller_;
-  position_controller::PositionControllerParams base_controller_params_;
+  Tcontroller base_controller_;
```
Comment (mfaessle) [inline control/autopilot/include/autopilot/autopilot.h]: I named this base controller because I thought that there might be new controllers that are only used for certain tasks but some of the basic tasks are always executed with the base controller (like start land etc.). @foehnph would something like this make sense (and then have two controller instances) or not?

**Comment by mfaessle** [comment]
About transforming the cpp file into a header... I never used templates myself but maybe @tcies has some comments about naming conventions etc of such files.

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 1 above)
Comment (kohlerj) [reply control/autopilot/include/autopilot/autopilot_inl.h]: There you go, the reason I was searching for! Make sense to me

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 3 above)
Comment (kohlerj) [reply control/autopilot/include/autopilot/autopilot_inl.h]: Conventions are here to be respected. Ok

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 4 above)
Comment (kohlerj) [reply control/autopilot/include/autopilot/autopilot_inl.h]: @foehnph Agreed for the 3rd bullet, you are right!

File: control/autopilot/include/autopilot/autopilot.h (hunk 7 above)
Comment (kohlerj) [reply control/autopilot/include/autopilot/autopilot.h]: I agree with @mfaessle, it makes it easier for you @foehnph since you don't need to implement the velocity mode in mpc, no?

File: control/autopilot/src/autopilot_node.cpp [hunk 8]
```diff
@@ -3,7 +3,8 @@
 int main(int argc, char **argv)
 {
   ros::init(argc, argv, "autopilot");
-  autopilot::AutoPilot autopilot;
+  autopilot::AutoPilot<position_controller::PositionController,
```
[the remark this answers is not in the document]
Comment (foehnph) [reply control/autopilot/src/autopilot_node.cpp]: done

File: control/autopilot/include/autopilot/autopilot.h [hunk 9]
```diff
@@ -228,3 +231,5 @@ class AutoPilot
 };
 
 } // namespace autopilot
+
+#include "./autopilot_inl.h"
```
[the remark this answers is not in the document]
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot.h]: there is none... do you mean before the inline include?

[the remark this answers is not in the document]
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot.h]: done

File: control/autopilot/parameters/default.yaml [hunk 10]
```diff
@@ -27,3 +27,4 @@ emergency_land_thrust: 9.0 # [m/s^2]
 
 control_command_input_timeout: 0.1 # [s]
 enable_command_feedthrough: false
+predictive_control_lookahead: 2.0 # [s]
```
[the remark this answers is not in the document]
Comment (foehnph) [reply control/autopilot/parameters/default.yaml]: done

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 5 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: Yes but the you do not guarantee that you reach the end state...
the end of a trajectory normally has zero velocity ;-) and even if not it just means that you get to a stop with an offset where the velocity error is compensated by the position error.

I propose to change `setAutoPilotStateForced()` so that it sets reference state velocity to zero when trigger to hover.

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 6 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: we use two spaces for tabs... so if I see it I change it :-)

File: control/autopilot/include/autopilot/autopilot.h (hunk 7 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot.h]: not really... then you would have to switch between controllers, implement state machine for that and generally reimplement a shitload of things if you want to use the autopilot with the new controller and not just have a different controller in feedthrough...

**Comment by foehnph** [comment]
@mfaessle the `autopilot_inl.h` is done according to @tcies proposed convention

File: control/autopilot/parameters/default.yaml [hunk 11]
```diff
@@ -6,7 +6,7 @@ start_land_velocity: 0.5 # [m/s]
 start_idle_duration: 2.0 # [s]
 idle_thrust: 2.0 # [m/s]
 optitrack_start_height: 1.0 # [m]
-optitrack_start_land_timeout: 5 # [s]
+optitrack_start_land_timeout: 6 # [s]
```
[the remark this answers is not in the document]
Comment (foehnph) [reply control/autopilot/parameters/default.yaml]: sorry was testing... undone

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 1 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: OK, clarification:
When using the MPC, it lags behind since your take-off maneuver has a rising position, fixed velocity but zero acceleration and hover thrust, which is unfeasible at the beginning of the start maneuver.
This is no problem for feasible control with the MPC, but it means that the it lags behind due to the compromise of positon-, velocity-, acceleration- and thrust-error.

Possible solution:
1. Ignore ~0.15m z-offset after start.
2. Add a takeoff acceleration (which could be a parameter) at the beginning of the start maneuver.

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 2 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: done in https://github.com/uzh-rpg/rpg_quadrotor_common/pull/11

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 3 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: agree, mark as done

File: control/autopilot/include/autopilot/autopilot_inl.h (hunk 4 above)
Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: Ok temporary solution:
I commented the code there to make it understandable what it does.
I think this should be a separate class, like `autopilot::Tracker`, the trajectory queue could be passed as reference and it can have time-based tracking and spacial tracking, as well as restrictions to forward-only tracking (can speed up iterations massively).
Additionally it could also handle the trajectory queue if wanted, but I guess it's easier to leave that in the autopilot.

**Comment by foehnph** [comment]
updated rpg_quadrotor_common with new Trajectory constructor, so...
@rpg-jenkins test this please

**Comment by foehnph** [comment]
can someone approve this please?

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 12]
```diff
@@ -877,16 +899,27 @@ quadrotor_common::ControlCommand AutoPilot::start(
           + start_land_velocity_
               * (timeInCurrentState() - start_idle_duration_);
       reference_state_.velocity.z() = start_land_velocity_;
+      if(timeInCurrentState() < start_idle_duration_ +
```
Comment (mfaessle) [inline control/autopilot/include/autopilot/autopilot_inl.h]: This is a hack since the velocity does not change with acceleration over time. But as long as it works for both controllers I can live with it ;) The nicest way would be if it would be a real trapezoidal velocity profile during the start maneuver.

Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: I'm gonna change that, test in simulation and request an approve again

Comment (mfaessle) [reply control/autopilot/include/autopilot/autopilot_inl.h]: I opened an issue regarding this (#64). Feel free to adress it in a new pull request.

Comment (foehnph) [reply control/autopilot/include/autopilot/autopilot_inl.h]: already done here

File: control/autopilot/include/autopilot/autopilot_inl.h [hunk 13]
```diff
@@ -877,16 +899,29 @@ quadrotor_common::ControlCommand AutoPilot::start(
...
+         start_land_velocity_/start_land_acceleration_)
```
Comment (mfaessle) [inline control/autopilot/include/autopilot/autopilot_inl.h]: spaces around arithmetic symbols please

**Comment by foehnph** [comment]
finally happy? I'll merge tomorrow morning... ;-)
